libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,380 @@
+// API.
+type diskBlockCacheConfig interface {
+	codecGetter
+	logMaker
+	clockGetter
+}
+
+// DiskBlockCacheStandard is the standard implementation for DiskBlockCache.
+type DiskBlockCacheStandard struct {
+	config   diskBlockCacheConfig
+	maxBytes uint64
+	log      logger.Logger
+	// This protects the disk caches from being shutdown while they're being
+	// accessed.
+	lock    sync.RWMutex
+	blockDb *leveldb.DB
+	lruDb   *leveldb.DB
+}
+
+var _ DiskBlockCache = (*DiskBlockCacheStandard)(nil)
+
+// openLevelDB opens or recovers a leveldb.DB with a passed-in storage.Storage
+// as its underlying storage layer.
+func openLevelDB(stor storage.Storage) (db *leveldb.DB, err error) {
+	db, err = leveldb.Open(stor, leveldbOptions)
+	if _, isErrCorrupted := err.(*storage.ErrCorrupted); isErrCorrupted {
+		// There's a possibility that if the leveldb wasn't closed properly
+		// last time while it was being written, then the manifest is corrupt.
+		// This means leveldb must rebuild its manifest, which takes longer
+		// than a simple `Open`.
+		// TODO: log here
+		return leveldb.Recover(stor, leveldbOptions)
+	}
+	return db, err
+}
+
+// newDiskBlockCacheStandardFromStorage creates a new *DiskBlockCacheStandard
+// with the passed-in storage.Storage interfaces as storage layers for each
+// cache.
+func newDiskBlockCacheStandardFromStorage(config diskBlockCacheConfig,
+	blockStorage, lruStorage storage.Storage, maxBytes uint64) (
+	*DiskBlockCacheStandard, error) {
+	log := config.MakeLogger("KBC")
+	blockDb, err := openLevelDB(blockStorage)
+	if err != nil {
+		return nil, err
+	}
+
+	lruDb, err := openLevelDB(lruStorage)
+	if err != nil {
+		blockDb.Close()
+		return nil, err
+	}
+	return &DiskBlockCacheStandard{
+		config:   config,
+		maxBytes: maxBytes,
+		log:      log,
+		blockDb:  blockDb,
+		lruDb:    lruDb,
+	}, nil
+}
+
+func versionPathFromVersion(dirPath string, version uint64) string {
+	return filepath.Join(dirPath, fmt.Sprintf("v%d", version))
+}
+
+func getVersionedPathForDiskCache(dirPath string) (versionedDirPath string,
+	err error) {
+	// Read the version file
+	versionFilepath := filepath.Join(dirPath, versionFilename)
+	versionBytes, err := ioutil.ReadFile(versionFilepath)
+	// We expect the file to open successfully or not exist. Anything else is a
+	// problem.
+	version := initialDiskCacheVersion
+	if ioutil.IsNotExist(err) {
+		// Do nothing, meaning that we will create the version file below.
+	} else if err != nil {
+		return "", err
+	} else {
+		// We expect a successfully opened version file to parse a single unsigned
+		// integer representing the version. Anything else is a corrupted version
+		// file. However, this we can solve by deleting everything in the cache.
+		version, err = strconv.ParseUint(string(versionBytes), 10, strconv.IntSize)
+		if err != nil {
+			return "", err
+		}
+		if version < initialDiskCacheVersion {
+			return "", errors.WithStack(
```

Comment:
The comment on `OutdatedVersionError` seems to imply that it is the right one to use there, but I don't really care either way. I just want sane errors :P